src/Native/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -810,31 +810,87 @@ extern "C" void RaiseFailFastException(PEXCEPTION_RECORD arg1, PCONTEXT arg2, UI
     abort();
 }
 
-typedef void(__stdcall *PFLS_CALLBACK_FUNCTION) (void* lpFlsData);
+// Key for the thread local storage of the attached thread pointer
+static pthread_key_t g_threadKey;
 
-extern "C" UInt32 FlsAlloc(PFLS_CALLBACK_FUNCTION arg1)
+// Callback to call when a thread is shut down
+static ThreadShutdownCallback g_threadShutdownCallback = NULL;
+
+// Destructor of the thread local object represented by the g_threadKey,
+// called when a thread is shut down
+void TlsObjectDestructor(void* data)
 {
-    // UNIXTODO: The FLS stuff needs to be abstracted, the only usage of the FLS is to get a callback
-    // when a thread is terminating.
-    return 0;
+    UNREFERENCED_PARAMETER(data);
+    ASSERT(data == pthread_getspecific(g_threadKey));
+
+    g_threadShutdownCallback();
 }
 
-__thread void* flsValue;
+// Register a callback to call at each thread shutdown
+// Parameters:
+//  callback - the shutdown callback function
+// Return:
+//  true if the registration succeeded, false if it has failed 
+extern "C" bool PalRegisterThreadShutdownCallback(ThreadShutdownCallback callback)
+{
+    ASSERT(g_threadShutdownCallback == NULL);
+    g_threadShutdownCallback = callback;
+    int status = pthread_key_create(&g_threadKey, TlsObjectDestructor);
+
+    return (status == 0);
+}
 
-extern "C" void * FlsGetValue(UInt32 arg1)
+// Attach thread to PAL. 
+// It can be called multiple times for the same thread.
+// It fails fast if a different thread was already registered.
+// Parameters:
+//  thread        - thread to attach
+//  isFirstAttach - true if this is the first attach for the given thread
+// Return:
+//  true if the thread was not attached to PAL, false if it was
+extern "C" bool PalAttachThread(void* thread, bool isFirstAttach)
```

Comment:
The `isFirstAttach` argument looks like unnecessary complexity as you have mentioned in the comments ... it would be nice to get rid of it.